binaries/cuprated/src/rpc/json.rs

```diff
@@ -83,212 +125,1126 @@ pub(super) async fn map_request(
+        let mut aux_pow_raw = Vec::<[u8; 32]>::with_capacity(len);
+
+        assert_eq!(
+            aux_pow.len(),
+            slots.len(),
+            "these need to be the same or else the below .zip() doesn't make sense"
+        );
+
+        for (aux_pow, slot) in aux_pow.iter().zip(&slots) {
+            if u32_to_usize(*slot) >= len {
+                return Err(anyhow!("Slot value out of range"));
+            }
+
+            aux_pow_id_raw.push(aux_pow.id);
+            aux_pow_raw.push(aux_pow.hash);
+        }
+
+        assert_eq!(
+            slots.len(),
+            aux_pow_raw.len(),
+            "these need to be the same or else the below .zip() doesn't make sense"
+        );
+        assert_eq!(
+            aux_pow_raw.len(),
+            aux_pow_id_raw.len(),
+            "these need to be the same or else the below .zip() doesn't make sense"
+        );
+
+        for (slot, aux_pow) in slots.iter().zip(&aux_pow) {
+            let slot = u32_to_usize(*slot);
+
+            if slot >= len {
+                return Err(anyhow!("Slot value out of range"));
+            }
+
+            aux_pow_raw[slot] = aux_pow.hash;
+            aux_pow_id_raw[slot] = aux_pow.id;
+        }
+
+        (
+            aux_pow_id_raw.into_boxed_slice(),
+            aux_pow_raw.into_boxed_slice(),
+        )
+    };
+
+    fn tree_hash(aux_pow_raw: &[[u8; 32]]) -> [u8; 32] {
+        todo!("https://github.com/serai-dex/serai/pull/629")
+    }
+
+    fn encode_mm_depth(aux_pow_len: usize, nonce: u32) -> u64 {
+        todo!("https://github.com/monero-project/monero/blob/893916ad091a92e765ce3241b94e706ad012b62a/src/cryptonote_basic/merge_mining.cpp#L74")
+    }
+
+    let merkle_root = tree_hash(aux_pow_raw.as_ref());
+    let merkle_tree_depth = encode_mm_depth(len, nonce);
+
+    let block_template = {
+        let hex = hex::decode(request.blocktemplate_blob)?;
+        Block::read(&mut hex.as_slice())?
+    };
+
+    fn remove_field_from_tx_extra() -> Result<(), ()> {
+        todo!("https://github.com/monero-project/monero/blob/master/src/cryptonote_basic/cryptonote_format_utils.cpp#L767")
+    }
+
+    if remove_field_from_tx_extra().is_err() {
+        return Err(anyhow!("Error removing existing merkle root"));
+    }
+
+    fn add_mm_merkle_root_to_tx_extra() -> Result<(), ()> {
+        todo!("https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server.cpp#L2189
+        ")
+    }
+
+    if add_mm_merkle_root_to_tx_extra().is_err() {
+        return Err(anyhow!("Error adding merkle root"));
+    }
+
+    fn invalidate_hashes() {
+        // block_template.invalidate_hashes();
+        // block_template.miner_tx.invalidate_hashes();
+        // <https://github.com/monero-project/monero/blob/cc73fe71162d564ffda8e549b79a350bca53c454/src/rpc/core_rpc_server.cpp#L2195-L2196>
+        todo!();
+    }
+
+    invalidate_hashes();
```

Comment:
This section has some missing functions, any suggestions for where or if these should be implemented?